googlefirestore/src/main/java/com/yahoo/ycsb/db/firestore/GoogleFirestoreClient.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2019 YCSB contributors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package com.yahoo.ycsb.db.firestore;
+
+import com.google.api.core.ApiFuture;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.cloud.firestore.Query;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.SetOptions;
+import com.google.cloud.firestore.WriteResult;
+import com.yahoo.ycsb.ByteIterator;
+import com.yahoo.ycsb.DB;
+import com.yahoo.ycsb.DBException;
+import com.yahoo.ycsb.Status;
+import com.yahoo.ycsb.StringByteIterator;
+
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.Vector;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * YCSB Client for Google's Cloud Firestore.
+ */
+public class GoogleFirestoreClient extends DB {
+  /**
+   * The names of properties which can be specified in the config files and flags.
+   */
+  public static final class GoogleFirestoreProperties {
+    private GoogleFirestoreProperties() {}
+
+    /**
+     * The Cloud Firestore project ID to use when running the YCSB benchmark.
+     */
+    static final String PROJECT = "googlefirestore.projectId";
+    static final String PRIVKEYFILE = "googlefirestore.serviceAccountKey";
+    static final String DEBUG = "googlefirestore.debug";
+  }
+
+  private static final Logger LOGGER = Logger.getLogger(GoogleFirestoreClient.class);
+
+  private static Firestore fsDb;
+
+  @Override
+  public void init() throws DBException {
```

Comment:
YCSB will create a DB instance per client thread and call init on each of them.

The current implementation here has unguarded access to the static member fsDb. Instead either fsDb needs to be per-instance or you need to guard the concurrent access.

Most of the other bindings do some combination of synchronization and ref counting on a static member.